src/vos/vos_pool.c

```diff
@@ -390,16 +390,18 @@ vos_pool_create(const char *path, uuid_t uuid, daos_size_t scm_sz,
 		goto close;
 
 	/* Create a VOS pool handle using ph. */
-	rc = pool_open(ph, pool_df, uuid, flags, poh);
-	if (rc != 0)
+	rc_pool_open = pool_open(ph, pool_df, uuid, flags, poh);
+	if (rc_pool_open != 0)
 		goto close;
 	ph = NULL;
 
 close:
-	/* Close this local handle, if it hasn't been consumed by pool_open. */
-	if (ph != NULL)
+	/* Close this local handle, if it hasn't been consumed by pool_open
```

Comment:
Just trying to understand, no change suggested (other than maybe modifying the comments here and in pool_open() to be clearer):

I am having trouble understanding the existing comment "consumed by pool_open". does it mean that in the rc_pool_open == 0 case that (somehow) a reference is taken that then makes it safe here to call vos_pmemobj_close() / pmemobj_close()? In pool_open() flow I don't see where that happens, e.g., cannot find a corresponding vos_pmemobj_open() / pmemobj_open().

Reply:
In fact it is a bit tricky, if rc_pool_open == 0 it has 2 different meanings, 1) pool_open() has never been called and ph must be closed/freed, 2) pool_open() has been called and was successful so that the remaining open process has completed and thus ph must be kept, so this is why it is NULL'ed ! if rc_pool_open != 0, pool_open() processing has encountered a failure and vos_pool_decref()/.../pool_hop_free()/vos_pmemobj_close() has already been called, and thus ph has already been closed/freed.

Reply:
It's likely my bad. From my perspective, the bug is that pool_open should not close ph. Could we simply fix pool_open so that it never closes ph? E.g., by replacing the vos_pool_decref(pool) call with explicit resource release operations, which do not include closing ph?

Reply:
> Could we simply fix pool_open so that it never closes ph? E.g., by replacing the vos_pool_decref(pool) call with explicit resource release operations, which do not include closing ph?

Hmmm, it looks a bit tricky to ensure that all but pmemobj_close() of the underlying actions implied by a call to vos_pool_decref() for the last reference ! By the way I had unsuccessfully tried what I thought to be the simplest way (ie, avoid calling vos_pmemobj_close() in pool_hop_free() ...) with the first commit for this PR.

Reply:
Like this:

    diff --git a/src/vos/vos_pool.c b/src/vos/vos_pool.c
    index a0503a806..cd4f831c8 100644
    --- a/src/vos/vos_pool.c
    +++ b/src/vos/vos_pool.c
    @@ -706,8 +706,16 @@ pool_open(PMEMobjpool *ph, struct vos_pool_df *pool_df, uuid_t uuid,
            gc_add_pool(pool);
            D_DEBUG(DB_MGMT, "Opened pool %p\n", pool);
            return 0;
    +
     failed:
    -       vos_pool_decref(pool); /* -1 for myself */
    +       if (pool->vp_io_ctxt != NULL)
    +               bio_ioctxt_close(pool->vp_io_ctxt, pool->vp_pool_df->pd_nvme_sz == 0);
    +       if (pool->vp_vea_info != NULL)
    +               vea_unload(pool->vp_vea_info);
    +       if (daos_handle_is_valid(pool->vp_cont_th))
    +               dbtree_close(pool->vp_cont_th);
    +       vos_dedup_fini(pool);
    +       D_FREE(pool);
            return rc;
     }
     
(I think it's not a good idea anyway to destruct a partially constructed object using a destructor meant for a completely constructed object.)

Alternatively, make pool_open to close the PMDK pool upon _any_ error (I don't think this is a good semantic though):

    diff --git a/src/vos/vos_pool.c b/src/vos/vos_pool.c
    index a0503a806..a6a67a2e7 100644
    --- a/src/vos/vos_pool.c
    +++ b/src/vos/vos_pool.c
    @@ -391,8 +391,6 @@ open:
     
            /* Create a VOS pool handle using ph. */
            rc = pool_open(ph, pool_df, uuid, flags, poh);
    -       if (rc != 0)
    -               goto close;
            ph = NULL;
     
     close:
    @@ -611,8 +609,8 @@ vos_register_slabs(struct umem_attr *uma)
     }
     
     /*
    - * If successful, this function consumes ph, which the caller shall not close
    - * in this case.
    + * If successful, this function consumes ph; upon any error, this function
    + * closes ph.
      */
     static int
     pool_open(PMEMobjpool *ph, struct vos_pool_df *pool_df, uuid_t uuid,
    @@ -628,6 +626,7 @@ pool_open(PMEMobjpool *ph, struct vos_pool_df *pool_df, uuid_t uuid,
            rc = pool_alloc(uuid, &pool); /* returned with refcount=1 */
            if (rc != 0) {
                    D_ERROR("Error allocating pool handle\n");
    +               vos_pmemobj_close(ph);
                    return rc;
            }
     
    @@ -788,8 +787,6 @@ vos_pool_open(const char *path, uuid_t uuid, unsigned int flags,
            }
     
            rc = pool_open(ph, pool_df, uuid, flags, poh);
    -       if (rc != 0)
    -               goto out;
            ph = NULL;
     
     out:

Reply:
@liw , ok I have pushed a new/3rd commit to implement this 2nd solution (to "make pool_open to close the PMDK pool upon any error") that you have proposed as a better way to fix than my 2nd commit for this PR.

Reply:
Thanks for considering this, Bruno.